### ISSUE TITLE
Failover mechanism improvement async client

### ISSUE THREAD

**Issue by hnotty** [eclipse-paho/paho.mqtt.c#989]
This might actually be a bug rather than a feature that needs to be added. However I couldn't find what I was after in the source code. This is for the Async client. I haven't used the other clients so I do not know if it is applicable to them or not.

A very useful feature of the failover mechanism would to be able to parse the serverURI list continuously. At the moment, the code parses the list only once. Once it has reached the end of the list it stays there and doesn't go back to the start of the list.
So basically it only tries the server addresses in the list once.
This feature is currently implemented in the Golang version of the Paho lib, so it would be nice to have it in the C one as well.

By adding the following lines at line 2196 in MQTTAsync.c, MQTTAsync_processCommand() the feature seems to works:

		} else if (command->command.type == CONNECT)
		{ /* Reached the point where we need to go back to the beginning of the serverURI list */
			Log(TRACE_MIN, -1, "Connect failed, retrying start of serverURI list");

			if (command->client->c->MQTTVersion == MQTTVERSION_DEFAULT)
			{
				if (command->command.details.conn.MQTTVersion == MQTTVERSION_3_1)
				{
					command->command.details.conn.currentURI = 0;
					command->command.details.conn.MQTTVersion = 	MQTTVERSION_DEFAULT;
				}
			} else
				command->command.details.conn.currentURI = 0; /* Here currentURI becomes larger than command->client->serverURIcount. This needs to be handled to avoid segmentation faults! */

			/* put the connect command back to the head of the command queue, using the next serverURI */
			rc = MQTTAsync_addCommand(command,
					sizeof(command->command.details.conn));
		} else

This feature is necessary when using AmazonMQ brokers which have failovers enabled. The second server for the AmazonMQ broker only seems to come online when the first one is down. When the first one is operational once again, server two is turned off and one turned back on. It's this last case the Paho C lib doesn't handle. As I said above the Go library does handle this.

**Comment by icraggs** [comment]
Does the connect ever fail in this case?  It's supposed to when it gets to the end of the list.

**Comment by hnotty** [comment]
@icraggs Thanks for the swift reply.
It's this you are referring to right ?
`5, 20201030 144809.491 getaddrinfo failed for addr xxxxxxxxxxxxxxx-2.mq.eu-central-1.amazonaws.com with rc -2`
`5, 20201030 144809.491 xxxxxxxxxxxxx-2.mq.eu-central-1.amazonaws.com is not a valid IP address`
`3, 20201030 144809.491 (3023041552)    (3)< Socket_new:845 (-2)`
`3, 20201030 144809.491 (3023041552)   (2)< MQTTProtocol_connect:324 (-2)`
`3, 20201030 144809.491 (3023041552)   (2)> MQTTAsync_disconnect1:3879`
`3, 20201030 144812.521 (3023041552)   (2)< MQTTAsync_disconnect1:3920 (-3)`
`3, 20201030 144812.521 (3023041552)   (2)> MQTTAsync_checkConn:567`
`3, 20201030 144812.521 (3023041552)   (2)< MQTTAsync_checkConn:570 (0)`

**Comment by hnotty** [comment]
I've noticed that my little fix has not covered all cases.

Doesn't seem to do anything if the client cannot connect to any servers from the start. Stays stuck on the last server of the list in that particular case.

**Comment by icraggs** [comment]
The sequence should be:

1. Call connect, successfully connect to one server in the list.  If none are available the connect fails.
2. If connection fails, automatically reconnect (if using), or call connect again to reconnect to any of the available servers in the list.  Can be a new list if you are calling connect explicitly.  The connect operation starts every time with the first URI in the list.

**Comment by hnotty** [comment]
Ok. AutoReconnect is enabled in my code.
And I've got the Connect() in the connlost callback.

**Comment by icraggs** [comment]
Those will be conflicting,  You should use one or the other.

**Comment by icraggs** [comment]
You can use the reconnect call, as opposed to a connect call, to start an auto-reconnect operation if it's in it's waiting period.

**Comment by icraggs** [comment]
So I think what you need should work, and if it doesn't it's a bug.  But it shouldn't work by cycling back to the front of the list.  It should work by a new connect, or reconnect, starting at the beginning of the list again.

**Comment by hnotty** [comment]
So basically it doesn't need to cycle back to the front of the list. So what I'm asking for is already there. That's what I thought.

It must be something I have not done properly then. I'll try what you've suggested and let you know the outcome.
At the moment I'm just using Connect() at startup with AutoReconnect set, no Connect() or Reconnect () in the conn_lost callback, and letting the client sort itself out.

**Comment by icraggs** [comment]
In auto reconnect mode the onsuccess or onfailure are only called once for the initial connect command.  There is a separate "connected" callback which is called when auto reconnect is successful. It's the counterpart to connectionLost.
